blog/_src/posts/2017-06-05-syntactic-parametricity-strikes-again.md

```diff
@@ -0,0 +1,399 @@
+by looking at the structure of canonical derivations. More precisely,
+I prove that
+\\(
+  ∀α. \List α → \List \alpha
+\\)
+is isomorphic to
+\\(
+    Π(n:\Nat{}). \List{(\Fin{n})}
+\\).
+
+<!-- more -->
+
+Context: Last week I had the pleasure of visiting UPenn, where I had
+many interesting discussions with various people. It was also an
+occasion to temporarily resume a discussion/collaboration I have with
+Li-Yao Xia, who is currently an intern there, and Jean-Philippe
+Bernardy, about testing polymorphic programs and its relation to
+canonical representations for System F.
+
+During one of our blackboard discussion, Li-Yao and I did a manual
+proof of a cool result: we proved a parametricity theorem for
+\\(∀α. \List α → \List α\\) using purely "syntactic" methods, using
+proof search among canonical proofs. (This is an idea that I have been
+playing with since the last year of my [PhD
+thesis](http://www.ccs.neu.edu/home/gasche/phd_thesis/), where
+I unsuccessfully tried to extend my work on canonical forms for the
+simply-typed lambda-calculus to polymorphism. It is here worked out on
+an specific example, but my end goal is to turn the manual reasoning
+into an algorithm.)
+
+You may wonder, first, why the isomorphism holds. The idea is that
+a polymorphic function of type \\(\List α → \List α\\) cannot inspect
+the elements of the input list; it can only use them in the resulting
+list, possibly duplicating, reordering or dropping some elements. On
+any input list of size \\(n\\), the behavior of the function can be
+described by a list of indices in \\([0; n-1]\\). For example, if the
+input \\([x, y, z]\\) (for some values of \\(x, y, z\\)) gives the
+output \\([y, y, x]\\), then this relation will hold on *any* value of
+\\(x, y, z\\), as the function cannot inspect their value or even test
+them for equality. The behavior of this function on lists of size
+3 can be fully described by the list of indices \\([1, 1, 0]\\). Its
+whole behavior is then uniquely determined by one such list for each
+possible size:
+
+\\[
+    ∀α. \List α → \List α  \quad≃\quad  Π(n:\Nat{}). \List{(\Fin n)}
+\\]
+
+The idea behind the "syntactic" (proof-theoretic?) proof method is the
+following: the set of closed values at a type \\(A\\) is isomorphic to
+the *search space* for canonical/normal derivations of
+\\(\judge{}{A}\\). We have tools (in particular the notion of
+*invertible* inference rule) to reason on those – in this post I will
+only present the reasoning informally, but it can easily be made
+formally precise.
+
+We start by looking at the shape of the search space for
+
+\\[
+    \judge{}{∀α. \List α → \List α}
+\\]
+or, said otherwise, of the judgment
+\\[
+    \judge{}{\List α → \List α}
+\\]
+
+with a fresh/abstract type variable \\(α\\). (I will not be keeping
+opened type variables in context to avoid confusing them with hypotheses.)
+
+Any derivation of a function type, without loss of generality
```

Comment:
Just found the phrase "wlog rules" at the end of the post.. okay. That's sad to me but I guess you do want to keep using wlog.